### PR TITLE
Observation analytics

### PR DESCRIPTION
**Pull request by mauberti-bc**
## Links to Jira Tickets
- N/A

## Description of Changes

- Adds an "Analytics" tab on the Survey page above the Observations data table
- Analytics endpoint for calculating summary statistics of observations for multiple Surveys
- Haven't written any unit tests in case the backend changes after review

### CODE REVIEW

File: api/src/paths/analytics/observations.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    if (!surveyIds || !Array.isArray(surveyIds) || surveyIds?.length === 0) {
+      throw new HTTP400('Invalid or missing surveyIds');
+    }
+
+    if (!groupByColumns || !Array.isArray(groupByColumns) || groupByColumns?.length === 0) {
+      throw new HTTP400('Invalid or missing groupByColumns');
+    }
+
+    const connection = getDBConnection(req['keycloak_token']);
```
Comment (MacDeluca) [inline api/src/paths/analytics/observations.ts]: You can use req.keycloak_token now

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/measurements/index.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ * @export
+ * @return {*}  {RequestHandler}
+ */
+export function getSurveyObservationMeasurements(): RequestHandler {
+  return async (req, res) => {
+    const surveyId = Number(req.params.surveyId);
+
+    defaultLog.debug({ label: 'getSurveyObservationMeasurement', surveyId });
+
+    const connection = getDBConnection(req['keycloak_token']);
```
Comment (MacDeluca) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/measurements/index.ts]: req.keycloak_token

File: api/src/paths/analytics/observations.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  }
+};
+
+export function getObservationCountByGroup(): RequestHandler {
+  defaultLog.debug({ label: 'getObservationCountByGroup' });
+
+  return async (req, res) => {
+    const { surveyIds, groupByColumns, groupByQuantitativeMeasurements, groupByQualitativeMeasurements } = req.query;
+
+    if (!surveyIds || !Array.isArray(surveyIds) || surveyIds?.length === 0) {
```
Comment (MacDeluca) [inline api/src/paths/analytics/observations.ts]: You can handle this logic in the openapi spec

File: api/src/models/observation-analytics.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  )
+});
+
+export type ObservationCountByGroupWithMeasurements = z.infer<typeof ObservationCountByGroupWithMeasurementsSchema>;
+
+export type ObservationCountByGroupSQLResponse = {
+  row_count: number;
+  individual_count: number;
+  individual_percentage: number;
+  [key: string]: any; // To accommodate dynamic columns
```
Comment (MacDeluca) [inline api/src/models/observation-analytics.ts]: Is this truly ```any``` or something like ```string | number | null | undefined"```

File: api/src/services/analytics-service.ts
Comment (MacDeluca) [inline api/src/services/analytics-service.ts:1]: Missing JSDoc for methods

**Review by NickPhura — DISMISSED** [review]
🐦

**Review by NickPhura — DISMISSED** [review]
🐦 🐦

File: api/src/models/observation-analytics.ts (hunk 4 above)
Comment (NickPhura) [reply api/src/models/observation-analytics.ts]: I was able to turn this interface into a zod schema by adding `catchall` (https://zod.dev/?id=catchall) to it to handle the dynamic fields. I had it set to string or number or null, but the upstream type handling blew up, so I had to set it to any.

There was some talk only about this issue, but no solution yet.

File: api/src/services/analytics-service.ts
Comment (NickPhura) [reply api/src/services/analytics-service.ts:1]: Added

**Review by NickPhura — APPROVED** [review]
🐦 🐦 🐦

**Pull request merged**